src/Blocks/BlockPiston.cpp

```diff
@@ -86,28 +73,129 @@ bool cBlockPistonHandler::GetPlacementBlockTypeMeta(
 
 
 
-int cBlockPistonHandler::FirstPassthroughBlock(int a_PistonX, int a_PistonY, int a_PistonZ, NIBBLETYPE pistonmeta, cWorld * a_World)
+Vector3i cBlockPistonHandler::VectorFromMetaData(int a_PistonMeta)
 {
-	// Examine each of the 12 blocks ahead of the piston:
-	for (int ret = 0; ret < PISTON_MAX_PUSH_DISTANCE; ret++)
+	switch (a_PistonMeta & 0x07)
 	{
-		BLOCKTYPE currBlock;
-		NIBBLETYPE currMeta;
-		AddPistonDir(a_PistonX, a_PistonY, a_PistonZ, pistonmeta, 1);
-		a_World->GetBlockTypeMeta(a_PistonX, a_PistonY, a_PistonZ, currBlock, currMeta);
-		if (cBlockInfo::IsPistonBreakable(currBlock))
+		case 0: return Vector3i( 0, -1,  0);
+		case 1: return Vector3i( 0,  1,  0);
+		case 2: return Vector3i( 0,  0, -1);
+		case 3: return Vector3i( 0,  0,  1);
+		case 4: return Vector3i(-1,  0,  0);
+		case 5: return Vector3i( 1,  0,  0);
+		default:
 		{
-			// This block breaks when pushed, extend up to here
-			return ret;
+			LOGWARNING("%s: invalid direction %d, ignoring", __FUNCTION__, a_PistonMeta & 0x07);
+			return Vector3i();
```

Comment:
I think adding an `ASSERT` would be a good choice here. We want any code calling this with bad parameters to fail hard in debug mode.